walkthroughs/week-5-tdd/project/src/main/java/com/google/sps/FindMeetingQuery.java

```diff
@@ -15,9 +15,110 @@
 package com.google.sps;
 
 import java.util.Collection;
+import java.util.Collections;
+import java.util.ArrayList;
+import java.util.List;
+import java.util.stream.Collectors;
+import java.util.stream.Stream;
+import java.lang.Math;
 
 public final class FindMeetingQuery {
+    private static final TimeRange START_OF_DAY_TIMERANGE = TimeRange.fromStartDuration(
+        TimeRange.START_OF_DAY, 0);
+    private static final TimeRange END_OF_DAY_TIMERANGE = TimeRange.fromStartDuration(
+        TimeRange.END_OF_DAY + 1, 0);
+
+  // returns time ranges that will accomadate meeting request
+  // if no time ranges exist all (mandatory + optional) attendees can attend, time ranges that
+  // accomadate only mandatory attendees is returned instead
   public Collection<TimeRange> query(Collection<Event> events, MeetingRequest request) {
-    throw new UnsupportedOperationException("TODO: Implement this method.");
+    Collection<TimeRange> mandatoryAttendeeTimeRanges = extractAttendingTimeRangesFromEvents(
+        events, request, /* only mandatory attendees */ false);
+    Collection<TimeRange> allAttendeeTimeRanges = extractAttendingTimeRangesFromEvents(
+        events, request, /* include optional attendees */ true);
+
+    Collection<TimeRange> acceptableMandatoryTimeRanges = getConflictFreeTimeRanges(
+        mandatoryAttendeeTimeRanges, request);
+    Collection<TimeRange> acceptableOptionalTimeRanges = getConflictFreeTimeRanges(
+        allAttendeeTimeRanges, request);
+
+    if (request.getAttendees().isEmpty() || !acceptableOptionalTimeRanges.isEmpty()) {
+      return acceptableOptionalTimeRanges;
+    } else {
+      return acceptableMandatoryTimeRanges;
+    }
```

Comment:
Changed it so that the mandatory time ranges are now instantiated/ found in the else clause. Thus, the extra instantiations/ logic only happens if no time ranges can be found that handle all attendees, optional or otherwise.